include/swift/AST/SILOptimizerRequests.h

```diff
@@ -87,6 +87,26 @@ class LoweredSILRequest
                                       ASTLoweringDescriptor desc) const;
 };
 
+class IsSelfRecursiveRequest
```

Comment:
I think using some of the existing "SIL analysis" infrastructure for caching this is actually more appropriate than the request evaluator, because SIL has a notion of invalidating an analysis when the function body changes (imagine it becomes self-recursive but only after several rounds of inlining).

Reply:
Right.
A SILFunction might change during the optimization pipeline. It might not be detected self-recursive earlier in the pipeline, e.g. because it calls a class method. After the method call is de-virtualized it becomes self-recursive.
Caching this with an analysis solves the problem, because if the function is changed, the analysis is invalidated and re-computed next time it is used.